### PR TITLE
[mlir][tensor][NFC] Improve documentation of pack/unpack

### PR DESCRIPTION
**Pull request by matthias-springer**

### CODE REVIEW

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td [hunk 1]
```diff
@@ -1751,42 +1751,66 @@ def Tensor_PackOp : Tensor_RelayoutOp<"pack", [
...
-    tiling factor for each dimension. The optional attribute `outer_dims_perm`
-    specifies the order for the tiled data dimension, while the attribute
-    `padding_value` specifies a padding value at the boundary on non-perfectly
-    divisible dimensions. Padding is optional:
-    - If absent, it is UB if the tile does not perfectly divide the dimension.
-    - If present, it will pad along high dimensions (high-padding) to make the
-      tile complete.
+    The "pack" operation converts a source tensor of rank `n` into a result
+    tensor of rank `n + k` with a tiled and packed layout (maybe with padding)
+    and optionally transposes some dimensions.
```
Comment (chelini) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:1756]: "some dimensions." I would clarify that we can only transpose the original tiled loop dimensions. We cannot interchange the new point loop dimensions.

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td [hunk 2]
```diff
@@ -1751,42 +1751,66 @@ def Tensor_PackOp : Tensor_RelayoutOp<"pack", [
...
+
+    Example:
     ```mlir
-    %0 = tensor.pack %source inner_dims_pos = [0, 1]
-      inner_tiles = [8, 32] into %dest : tensor<128x256xf32> -> tensor<16x8x8x32xf32>
+    // NC to NCnc
+    %0 = tensor.pack %source inner_dims_pos = [0, 1] inner_tiles = [8, 32]
+        into %dest : tensor<128x256xf32> -> tensor<16x8 x 8x32 xf32>
+    //                                             \  /   \  /
+    //                                       outer dims  inner dims
```
Comment (chelini) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:1779]: nice :)

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td [hunk 3]
```diff
@@ -1751,42 +1751,66 @@ def Tensor_PackOp : Tensor_RelayoutOp<"pack", [
...
+    being tiled, where `0 < k <= n`. The order of the dimensions matters: the
+    tiled dimensions (of size `inner_tiles`) are added to the end of the result
+    tensor in the order in which they appear in `inner_dims_pos`.
 
-    Example NC_to_NCnc:
+    `inner_tiles` (mandatory) specifies `k` tile sizes. These tile sizes
+    correspond to the least significant ("inner") result tensor dimension sizes.
+    Tile sizes can be static or dynamic.
 
+    Example: If `inner_tiles = [16, 32]`, the result tensor has shape
```
Comment (chelini) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:1767]: has a shape...

File: mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td [hunk 4]
```diff
@@ -1751,42 +1751,66 @@ def Tensor_PackOp : Tensor_RelayoutOp<"pack", [
...
 
+    Example:
     ```mlir
-    %0 = tensor.pack %arg padding_value(%pad : f32) inner_dims_pos = [0, 1]
-      inner_tiles = [8, 2] into %arg1 : tensor<13x15xf32> -> tensor<2x8x8x2xf32>
+    %0 = tensor.pack %arg0 padding_value(%pad : f32) outer_dims_perm = [2, 1, 0]
+        inner_dims_pos = [1] inner_tiles = [2] into %arg1
+        : tensor<199x127x256xf32> -> tensor<256x64x200x2xf32>
+    //            /    \
+    //  paddded dim   padded and tiled dim
```
Comment (chelini) [inline mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td:1807]: This was not the original intent. Padding should happen only on dimensions you tile. The verifier does not catch it and it is a bug.

**Review by chelini — COMMENTED** [review]
Thank you!

**Review by chelini — COMMENTED** [review]
Thanks, two more things and it is good to go.

**Review by nicolasvasilache — APPROVED** [review]
Thanks for cleaning this up, let's address @chelini 's comments and land this

**Pull request merged**